content/client-server-api/modules/secrets.md

```diff
@@ -66,6 +66,62 @@ default key.
 |------------|-----------|------------------------------------------|
 | key        | string    | **Required.** The ID of the default key. |
 
+
+###### `m.secret_storage.v1.aes-hmac-sha2`
+
+For the purposes of allowing clients to check whether a user has correctly
+entered the key, keys for use with the `m.secret_storage.v1.aes-hmac-sha2`
+algorithm are stored with some additional data.
+
+When storing a key, clients should:
+
+1.  Given the secret storage key, generate 64 bytes by performing an
+    HKDF with SHA-256 as the hash, a salt of 32 bytes of 0, and the empty
+    string as the info. The first 32 bytes are used as the AES key,
+    and the next 32 bytes are used as the MAC key.
+
+2.  Generate 16 random bytes, set bit 63 to 0 (in order to work around
+    differences in AES-CTR implementations), and use this as the AES
+    initialization vector (IV).
+
+3.  Encrypt a message consisting of 32 byutes of 0, using AES-CTR-256 using the
+    AES key and IV generated above.
+
+4.  Pass the raw encrypted data through HMAC-SHA-256 using the MAC key
+    generated above.
+
+5.  Encode the IV from step 2, and the MAC from step 4, using base64, and store
+    the results in the `iv` and `mac` properties respectively in the
+    `m.secret_storage.key.[key ID]` account-data. (The ciphertext from step 3
+    is discarded after passing through the MAC calculation.)
+
+This process can be repeated by a client checking if the key is correct: the
+MAC should match if the key is correct.
+
+The structure of a `m.secret_storage.key.[key ID]` account data object for use
+with this algorithm is therefore as follows:
+
+`AesHmacSha2KeyDescription`
+
+| Parameter   | Type   | Description                                                                                          |
+|-------------|--------|------------------------------------------------------------------------------------------------------|
+| name        | string | Optional. The name of the key.                                                                       |
+| algorithm   | string | **Required.** The encryption algorithm to be used for this key: `m.secret_storage.v1.aes-hmac-sha2`. |
+| passphrase  | object | See [deriving keys from passphrases](#deriving-keys-from-passphrases) section for a description of this property. |
+| iv          | string | The 16-byte initialization vector, encoded with base64.                                              |
+| mac         | string | The MAC of the result of encrypting 32 bytes of 0, encoded with base64.                              |
```

Comment:
Question: are these required or not? If not, what are clients expected to do if they are absent? Reject the key, or carry on regardless?

cf https://github.com/vector-im/element-web/issues/26721

Reply:
The [MSC said](https://github.com/matrix-org/matrix-spec-proposals/pull/2472/files#diff-c1213725eb7eea9b96e7fa6af61456682fb88a93640423c3c43727abd7faa96aR46-R52) that clients "should" calculate and store the IV and MAC, not "must", so I guess they're optional.  I don't remember why we made it optional.

If they're not present, we have to assume the key is correct, and carry on.